pkg/session/txn.go

```diff
@@ -652,6 +653,11 @@ func KeyNeedToLock(k, v []byte, flags kv.KeyFlags) bool {
 	}
 
 	if tablecodec.IsTempIndexKey(k) {
+		// We force DMLs to lock all temporary index keys in next-gen, because
+		// next-gen enforces conflict check on all keys, including non-unique index keys.
+		if kerneltype.IsNextGen() {
```

Comment:
Is there a test case to cover this path?

Reply:
Added. I'm not sure about whether the merge index case is reasonable. @tangenta @D3Hunter could you help take a look at it?